core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/keystore/persistence/KeyStoreRegistryPersistenceManager.java

```diff
@@ -0,0 +1,421 @@
+
+    /**
+     * Generates an ID for the public certificate, which is used as a file name suffix for the certificate.
+     * e.g. If keystore name is 'example-com.jks', public cert name will be 'example-com-343743.cert'.
+     *
+     * @return generated id to be used as a file name appender.
+     */
+    private String generatePublicCertId() {
+
+        String uuid = UUIDGenerator.getUUID();
+        return uuid.substring(uuid.length() - 6, uuid.length() - 1);
+    }
+
+    /**
+     * Get the key store from the registry.
+     *
+     * @param keyStoreName Name of the key store.
+     * @param tenantId     Tenant Id.
+     * @return Key store.
+     * @throws SecurityException If an error occurs while getting the key store.
+     */
+    public KeyStore getKeyStore(String keyStoreName, int tenantId) throws SecurityException {
+
+        Registry registry = getGovernanceRegistry(tenantId);
+        if (!isKeyStoreExistsInRegistry(keyStoreName, registry)) {
+            throw new SecurityException("Key Store with a name: " + keyStoreName + " does not exist.");
+        }
+
+        try {
+            String path = RegistryResources.SecurityManagement.KEY_STORES + REGISTRY_PATH_SEPARATOR + keyStoreName;
+            Resource resource = registry.get(path);
+            byte[] bytes = (byte[]) resource.getContent();
+            KeyStore keyStore = KeystoreUtils.getKeystoreInstance(resource
+                    .getProperty(RegistryResources.SecurityManagement.PROP_TYPE));
+            char[] passwordChar = getKeyStorePasswordFromRegistryResource(resource, keyStoreName);
+            ByteArrayInputStream stream = new ByteArrayInputStream(bytes);
+            keyStore.load(stream, passwordChar);
+            resource.discard();
+            return keyStore;
+        } catch (RegistryException | KeyStoreException | NoSuchProviderException | IOException | CertificateException |
+                 NoSuchAlgorithmException e) {
+            throw new SecurityException("Error getting key store: " + keyStoreName, e);
+        }
+    }
+
+    /**
+     * Method to retrieve list of keystore metadata of all the keystores in a tenant.
+     *
+     * @param tenantId Tenant Id.
+     * @return List of KeyStoreMetaData objects.
+     * @throws SecurityException If an error occurs while retrieving the keystore data.
+     */
+    public List<KeyStoreMetadata> listKeyStores(int tenantId) throws SecurityException {
+
+        Registry registry = getGovernanceRegistry(tenantId);
+        List<KeyStoreMetadata> metadataList = new ArrayList<>();
+        try {
+            if (!registry.resourceExists(RegistryResources.SecurityManagement.KEY_STORES)) {
+                return metadataList;
+            }
+            Collection keyStoreCollection = (Collection) registry.get(RegistryResources.SecurityManagement.KEY_STORES);
+            String[] keyStoreFullNames = keyStoreCollection.getChildren();
+
+            for (String keyStoreFullName : keyStoreFullNames) {
```

Comment:
Lets use streams.

Reply:
If we used streams, the code would look like this:

```
Arrays.stream(keyStorePaths)
                    .filter(keyStorePath -> !RegistryResources.SecurityManagement.PRIMARY_KEYSTORE_PHANTOM_RESOURCE.equals(
                            keyStorePath))
                    .forEach(keyStorePath -> {
                        try {
                            metadataList.add(getKeyStoreMetadata(keyStorePath, registry, tenantId));
                        } catch (RegistryException e) {
                            throw new SecurityException("Error when getting keyStore metadata.", e);
                        }
                    });

```
For readability IMO is using for loop is better than the streams. WDYT?

Reply:
+1